apps/helloshard/server.py

```diff
@@ -0,0 +1,422 @@
+            # a small modification to randousha
+            end_time = time.time()
+            logging.debug(
+                f"[{self.global_id}] Randousha finished in {end_time-start_time}"
+            )
+            logging.debug(f"len(rs_t): {len(rs_t)}")
+            logging.debug(f"rs_t: {rs_t}")
+            self._inputmasks += rs_t
+
+            # Step 1. III) Submit an updated report
+            await self._preprocess_report()
+
+            # Increment the preprocessing round and continue
+            preproc_round += 1
+
+    async def _client_request_loop(self):
+        # Task 2. Handling client input
+        # TODO: if a client requests a share,
+        # check if it is authorized and if so send it along
+        pass
+
+    def _collect_client_input(self, *, index, queue):
+        # Get the public input (masked message)
+        masked_message_bytes, inputmask_idx = queue(index)
+        masked_message = field(int.from_bytes(masked_message_bytes, "big"))
+        # Get the input mask
+        logging.debug(f"[{self.global_id}] inputmask idx: {inputmask_idx}")
+        logging.debug(f"[{self.global_id}] inputmasks: {self._inputmasks}")
+        try:
+            inputmask = self._inputmasks[inputmask_idx]
+        except KeyError as err:
+            logging.error(err)
+            logging.error(f"[{self.global_id}] inputmask idx: {inputmask_idx}")
+            logging.error(f"[{self.global_id}] inputmasks: {self._inputmasks}")
+
+        msg_field_elem = masked_message - inputmask
+        return msg_field_elem
+
+    # TODO generalize client and intershard collection
+    def _collect_intershard_msg(self, *, index, queue):
+        masked_message_bytes, mask_idx = queue(index)
+        masked_message = field(int.from_bytes(masked_message_bytes, "big"))
+        mask = self.intershardmask_shares[mask_idx]
+        msg_field_elem = masked_message - mask
+        return msg_field_elem
+
+    async def _mpc_loop(self):
+        # Task 3. Participating in MPC epochs
+        contract_concise = ConciseContract(self.contract)
+        n = contract_concise.n()
+        t = contract_concise.t()
+
+        epoch = 0
+        while True:
+            # 3.a. Wait for the next MPC to be initiated
+            while True:
+                epochs_initiated = contract_concise.epochs_initiated()
+                if epochs_initiated > epoch:
+                    break
+                await asyncio.sleep(5)
+
+            if self.is_gateway_shard:
+                client_msg_field_elem = self._collect_client_input(
+                    index=epoch, queue=contract_concise.input_queue
+                )
+
+                # 3.d. Call the MPC program
+                async def prog(ctx):
+                    logging.info(f"[{self.global_id}] Running MPC network")
+                    client_msg_share = ctx.Share(client_msg_field_elem)
+                    client_msg = await client_msg_share.open()
```

Comment:
This MPC program looks unclear. Which shard runs it, and also isn't clients secret misnamed?

Reply:
Yes this needs to be improved, Ideally it would be run by any shard that has received a client secret. In this code one shard acts as a gateway shard (hence the condition at line 251 above: `if self.is_gateway_shard:`) and only the gateway shard is expected to receive inputs from the client. Whether a server is part of a gateway shard or not is set at start up time in the `main` module.

I guess that the `_mpc_initiate_loop` logic needs to be adapted to shards. That is, each shard needs to have an `inputs_ready` condition that when met can be use to trigger the start of the client's input processing by the relevant shard.

Reply:
As for the name, yeah ... I now wonder what happened ... I think I was trying to be better understand what type each variable is and tried to pick names that could reflect this. Definitely can be changed!

It is indeed confusing. The method `_collect_client_input()` is used to read the client secret from the input queue but rather than just returning the client secret (`m̅`) it subtracts the input mask share (`[r]`) and returns the result which is a `field` element that is then cast to a `Share` (`[m]`). The naming was trying to reflect this I think ... :smile:.